API linter

roycaihw opened the pull request:
**UPDATE:**
The API convention "optional field should be a pointer" is controversial. We decided to not go with it for now. In this PR we added a simple validator that verifies ["Other than capitalization of the initial letter(/word), Go field name and JSON field name should almost always match"](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#naming-conventions) to make sure that the linter framework works. The "optional field should be a pointer" code and whitelist are stored [here](https://github.com/roycaihw/kubernetes/blob/c67214b10c2ab5df243866ccf62da1f418aff136/staging/src/k8s.io/code-generator/cmd/api-linter/linters/optionalfield.go), and can be submitted in another PR after this one gets merged.

--------------------------------------------------------------------------------------------

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Implements an API linter proposed in #44005. Currently we only have one policy "optional field should be a pointer" implemented. Policies suggested are tracked in #44005 and https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md. Current API violations are whitelist-ed to prevent the workflow from failing. 

**To run against all packages: (at root directory)**
`make generated_files`
`make verify_generated_files`
(The above commands won't fail with whitelist enabled. To disable it, modify `Makefile.generated_file` or make the `whitelist.csv` empty)

**Test commands:**
With whitelist:
`go run staging/src/k8s.io/code-generator/cmd/api-linter/main.go -i k8s.io/apimachinery/pkg/runtime --v 2 \`
` --alsologtostderr --whitelist-file staging/src/k8s.io/code-generator/cmd/api-linter/linters/whitelist.csv`

Without whitelist:
`go run staging/src/k8s.io/code-generator/cmd/api-linter/main.go -i k8s.io/apimachinery/pkg/runtime --v 2 --alsologtostderr`


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```
/sig api-machinery

smarterclayton commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/whitelist.csv:
```diff
@@ -0,0 +1,1019 @@
+optional_fields_linter_pointers,io.k8s.apimachinery.pkg.runtime.TypeMeta,APIVersion
```
I think this is highlighting how "should be a pointer" is silly and not delivering value.  As I look through this list, I wouldn't make most of these pointers.  Why have a linter that checks things that don't make sense?

mbohlool replied:
+ @bgrant0607

Are you suggesting we should change our [API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required) for optional fields?  I think that should be discussed in another thread (probably in sig-architecture). Meanwhile do you have any suggestion for another simpler linter that we can add here to make sure the linter framework works?

bgrant0607 replied:
First, our written conventions should reflect our actual conventions.

Second, I don't care how it's implemented, but we need to be able to uniformly distinguish set from unset fields, regardless of language-default values, and to omit unspecified fields when encoding resources that have not had defaults expanded. Pointers were a means to that end. If we can achieve it in a simpler manner, I'd be happy with that.

smarterclayton commented on build/root/Makefile.generated_files:
```diff
@@ -487,6 +488,108 @@ $(DEFAULTER_GEN):
 	touch $@
 
 #
+# Api linter
```
This make logic is so complex that I don't think it provides value (having worked with it in the past).  It's extremely difficult to debug and modify for minor performance gains.  Since in reality most users only see linting in PR jobs, I'd say this is something that should be removed and we should have a simpler invocation that uses normal processing of the packages.  goidl is fast at scanning all packages.

mbohlool replied:
I agree on the complexity of makefile. This can be (subjectively) simple scripts. However because the linter needs to work on the same package list as the other generators (specifically OpenAPI generator), do you think it make sense to duplicate that logic in two places? It would be nice (may be in another PR) if we can move all of these together, out of the Makefile into a bash script.

mbohlool commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/optionalfield.go:
```diff
@@ -0,0 +1,33 @@
+			if hasAPITagValue(m.CommentLines, tagValueFalse) {
+				continue
+			}
+			if hasOptionalTag(&m) && m.Type.Kind != types.Pointer {
```
With this implementation, maps and slices will be considered violation which is wrong. Our API convention is different from what we are checking here for maps and slices. The type need to be nillable (map, slice, or pointer to other types I guess).

lavalamp commented:
/assign @mml

mml commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/api-linter.go:
```diff
@@ -0,0 +1,245 @@
+	whitelistFilename string
+}
+
+type apiViolation struct {
```
Can you document any rules about setting these?  Like what is a suitable linterName or violationID?

Also, do we have access to file and line number where the type is declared?  This makes tracking down the failure a bit easier...  e.g., here's one from my conformance regtest.

```
ERROR at test/e2e/scheduling/predicates.go:245:2: Using It() with manual [Conformance] tag is no longer allowed.  Use framework.ConformanceIt() instead.
```

roycaihw replied:
Updated. Unfortunately we don't have access to file and line number in gengo.

mml replied:
Am I looking at the wrong file?  I don't see the update.

mml commented:
@roycaihw you need to run hack/update-bazel.

/ok-to-test

roycaihw commented:
> I would like to see us be able to build and run this as a test from bazel, but it is fine to do this as a followup. This is what I did with the conformance test regtest in #53909 and I'm happy to help. It will be easier to maintain than the Makefile version.

@mml We agreed to bazelify the make logic as a followup. It would be better to move all the generators' logic together to bazel. 

Other feedback addressed.

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw commented:
@mml I replaced the "optional field should be a pointer" validator with a "go_json_name_match" validator (see the updated first comment). This PR is ready for another review. PTAL.

roycaihw commented:
/retest

roycaihw commented:
/retest

as #55712 is merged

roycaihw commented:
/retest

roycaihw commented:
cross-build failure is #56002

roycaihw commented:
/retest

mml commented:
@roycaihw I will try to get to this again tomorrow.  Apologies for the delay.

roycaihw commented:
@mml No problem :)

mml commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/README:
```diff
@@ -0,0 +1,7 @@
+# API linter
```
Ping.

lavalamp replied:
Please make this file into a package comment instead of having it as a README.

mml commented on staging/src/k8s.io/code-generator/cmd/api-linter/main.go:
```diff
@@ -0,0 +1,45 @@
+	customArgs := &linters.CustomArgs{
+		WhitelistFilename: "",
+	}
+	pflag.CommandLine.StringVar(&customArgs.WhitelistFilename, "whitelist-file", customArgs.WhitelistFilename, "Specify a csv file that contains the whitelist.")
```
Ping.

mml commented:
Squash before submit.
/lgtm

roycaihw committed:
Use util.ToCanonicalName in kube-openapi

roycaihw committed:
Api linter framework and go-json-name-match validator

Run against all packages:
`make generated_files`
`make verify_generated_files`

Test commands:
go run staging/src/k8s.io/code-generator/cmd/api-linter/main.go -i
k8s.io/apimachinery/pkg/runtime --v 2 --alsologtostderr --whitelist-file
staging/src/k8s.io/code-generator/cmd/api-linter/linters/whitelist.csv
        (won't fail with whitelist provided)

Policies suggested are tracked at #44005

roycaihw commented:
@mml Squashed.

mml commented:
/lgtm

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw commented:
cc @lavalamp for approval

roycaihw commented:
/test pull-kubernetes-e2e-kubeadm-gce-canary

roycaihw commented:
seems to be a run_after_success behavior
/test all

krzyzacy commented:
the canary job does not blocking merge though, only `Required` contexts are needed :-)

BenTheElder commented:
> /test pull-kubernetes-e2e-kubeadm-gce-canary

These jobs (`.*-canary`) are used to validate configurations / images etc. in [test-infra](https://github.com/kubernetes/test-infra). Their results are relatively meaningless for other purposes.

lavalamp commented on staging/src/k8s.io/code-generator/cmd/api-linter/dummy/api_linter_dummy.go:
```diff
@@ -0,0 +1,19 @@
+limitations under the License.
+*/
+
+package dummy
```
What is this package for?

lavalamp commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/api-linter.go:
```diff
@@ -0,0 +1,258 @@
+	return "raw"
+}
+
+// Packages returns a API linter that validates API conventions and generates a dummy package.
```
Hopefully you can find a way of accomplishing this that doesn't actually involve making a new package.

roycaihw replied:
Currently `gengo` supports `assemble` and `verify` mode. I think to avoid generate new file/dir we need a third "dryrun" mode. I will open a PR for that

roycaihw commented:
@lavalamp Updated. It seems that `pull-kubernetes-verify` will complain if we don't commit the report file, because it runs `make verify_generated_files` first and then `hack/verify-godeps.sh` https://github.com/kubernetes/kubernetes/blob/d94bc9ef85b4e8e62f3deee2763d9195c9e7b4df/hack/jenkins/verify-dockerized.sh#L44
So it will detect an extra report file in `vendor/` directory.

roycaihw commented:
/remove-lifecycle stale

lavalamp commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/goJSONNameMatch.go:
```diff
@@ -0,0 +1,85 @@
+//     JSONSpec  jsonSpec    true
+//     JSONSpec  jsonspec    true
+
+package linters
```
I would just call this linter/file namesMatch. Maybe put each linter in their own package?

Maybe we shouldn't call these linters, maybe just "rules" is better. Linters sounds more optional than we want these to be.

lavalamp commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/goJSONNameMatch_test.go:
```diff
@@ -0,0 +1,40 @@
+		{"PodSpec", "spec", false},
+		{"Spec", "podSpec", false},
+		{"JSONSpec", "jsonSpec", true},
+		{"JSONSpec", "jsonspec", true},
```
This should be false.

lavalamp commented on staging/src/k8s.io/code-generator/cmd/api-linter/main.go:
```diff
@@ -0,0 +1,90 @@
+- To exclude a type or a member from a tagged package/type, add "+k8s:openapi-gen=false" tag to the comment lines.
+`
+
+func main() {
```
This file shouldn't exist when you move this logic over to kube-openapi

lavalamp reviewed (commented):
As discussed offline, let's move much of this logic to kube-openapi. I commented on things that I think you should do differently for your PR to that repo. While I'm out, you can get reviews from @mbohlool, @seans3, @apelisse, @pwittrock, and @erictune.

lavalamp commented:
...and sorry for the long delay...

roycaihw commented:
@lavalamp That's okay. Thanks for the feedback! I agree that the validation logic should be integrated into `kube-openapi` generator to provide a one stop shop experience, and much of the package filter and dummy package problems will be gone as we are actually generating something.

roycaihw commented on staging/src/k8s.io/code-generator/cmd/api-linter/linters/api-linter.go:
```diff
@@ -0,0 +1,292 @@
+
+func (l *apiLinter) Init(c *generator.Context, w io.Writer) error {
+	if l.whitelistFilename != "" {
+		glog.V(1).Infof("Reading whitelist file: %s", l.whitelistFilename)
```
The `generator.Context` is a global context for individual generators to consume. The whitelist / violations file is specific to the openAPIGen / apiLinter generator. IIUC the [generator construction](https://github.com/kubernetes/kube-openapi/blob/bf40560368791a7dddfeea9b3cfcf89b34139f44/pkg/generators/openapi.go#L129) happens after context is constructed. I can move the reading code to where [generator is constructed](https://github.com/kubernetes/kube-openapi/blob/bf40560368791a7dddfeea9b3cfcf89b34139f44/pkg/generators/openapi.go#L164) instead of when generator executes the `Init` method

roycaihw commented:
Closing in favor of https://github.com/kubernetes/kubernetes/pull/65737
